addons/binding/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/internal/UPBMessage.java

```diff
@@ -0,0 +1,309 @@
+         */
+        public static Command valueOf(byte value) {
+            return commandMap.get(value);
+        }
+    }
+
+    /**
+     * An enum of possible modem response types.
+     *
+     * @author Chris Van Orman
+     *
+     */
+    public enum Type {
+        ACCEPT,
+        BUSY,
+        ERROR,
+        ACK,
+        NAK,
+        MESSAGE_REPORT,
+        NONE;
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(UPBMessage.class);
+
+    private static HashMap<Integer, Command> commandMap = new HashMap<>();
+
+    static {
+        commandMap.put(0x20, Command.ACTIVATE);
+        commandMap.put(0x21, Command.DEACTIVATE);
+        commandMap.put(0x22, Command.GOTO);
+        commandMap.put(0x23, Command.START_FADE);
+        commandMap.put(0x24, Command.STOP_FADE);
+        commandMap.put(0x25, Command.BLINK);
+        commandMap.put(0x30, Command.REPORT_STATE);
+        commandMap.put(0x31, Command.STORE_STATE);
+        commandMap.put(0x86, Command.DEVICE_STATE);
+    }
+
+    private static HashMap<String, Type> typeMap = new HashMap<>();
+
+    static {
+        typeMap.put("PA", Type.ACCEPT);
+        typeMap.put("PB", Type.BUSY);
+        typeMap.put("PE", Type.ERROR);
+        typeMap.put("PK", Type.ACK);
+        typeMap.put("PN", Type.NAK);
+        typeMap.put("PU", Type.MESSAGE_REPORT);
+    }
+
+    /**
+     * Converts a hex string into a {@link UPBMessage}.
+     *
+     * @param commandString
+     *            the string as returned by the modem.
+     * @return a new UPBMessage.
+     */
+    public static UPBMessage fromString(String commandString) {
+        UPBMessage command = new UPBMessage();
+
+        String typeString = commandString.substring(0, 2);
+        Type type = Type.NONE;
+
+        if (typeMap.containsKey(typeString)) {
+            type = typeMap.get(typeString);
+        }
+
+        command.setType(type);
+
+        try {
+            if (commandString.length() > 2) {
+                byte[] data = DatatypeConverter.parseHexBinary(commandString.substring(2));
+                command.getControlWord().setBytes(data[1], data[0]);
+                int index = 2;
+                command.setNetwork(data[index++]);
+                command.setDestination(data[index++]);
+                command.setSource(data[index++]);
+
+                int commandCode = data[index++] & 0xFF;
+
+                if (commandMap.containsKey(commandCode)) {
+                    command.setCommand(commandMap.get(commandCode));
+                } else {
+                    command.setCommand(Command.NONE);
+                }
+
+                if (index < data.length - 1) {
+                    command.setArguments(Arrays.copyOfRange(data, index, data.length - 1));
+                }
+            }
+        } catch (Exception e) {
+            logger.error("Attempted to parse invalid message: {}", commandString, e);
+        }
+
+        return command;
+    }
+
+    private Type type;
```

Comment:
Please put all field definitions on top of the class, maybe you should extract the enums